### PR TITLE
Bug fix: Disable L059 in snowflake dialect

### LINKED ISSUE #3173: Snowflake - L059 corrupts SQL on column names created with quotes
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff fix creates unparseable sql if "date" is quoted

```
SELECT d."date" AS date_key
FROM foo.dates d
;
```
becomes
```
SELECT d.date AS date_key
FROM foo.dates d
;
```

Snowflake docs explicitly state date is not a reserved keyword, but it appears it actually is

### Expected Behaviour

-

### Observed Behaviour

-

### How to reproduce

-

### Dialect

Snowflake

### Version

0.13.0

### Configuration

-

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #3173 

Implements the solution discussed in the issue -- disabling this rule by default in Snowflake (and adding support for `force_enable`).

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/rules/L059.py [hunk 1]
```diff
@@ -87,6 +89,17 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
         self.prefer_quoted_identifiers: bool
         self.ignore_words: str
         self.ignore_words_regex: str
+        self.force_enable: bool
+        # Some dialects allow quotes as PART OF the column name. In other words,
+        # these are two different columns:
+        # - date
+        # - "date"
+        # For safety, disable this rule by default in those dialects.
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L059.py:97]: Can we add a `note` to docstring to explain this?

similar to this one: https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L026

Comment (barrywhart) [reply src/sqlfluff/rules/L059.py:104]: Done!

File: src/sqlfluff/rules/L059.py [hunk 2]
```diff
@@ -24,6 +24,13 @@ class Rule_L059(BaseRule):
     When ``prefer_quoted_identifiers = False`` (default behaviour), the quotes are
     unnecessary, except for reserved keywords and special characters in identifiers.
 
+    .. note::
+       This rule is disabled by default for Snowflake because it allow quotes as
```
Comment (WittierDinosaur) [inline src/sqlfluff/rules/L059.py:28]: typo
```suggestion
       This rule is disabled by default for Snowflake because it allows quotes as
```

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**